modules/reactor/src/meshgenerators/PinMeshGenerator.C

```diff
@@ -0,0 +1,480 @@
+  : MeshGenerator(parameters),
+    _reactor_params(getParam<MeshGeneratorName>("reactor_params")),
+    _pin_type(getParam<subdomain_id_type>("pin_type")),
+    _pitch(getParam<Real>("pitch")),
+    _num_sectors(getParam<unsigned int>("num_sectors")),
+    _ring_radii(isParamValid("ring_radii") ? getParam<std::vector<Real>>("ring_radii")
+                                           : std::vector<Real>()),
+    _duct_halfpitch(isParamValid("duct_halfpitch") ? getParam<std::vector<Real>>("duct_halfpitch")
+                                                   : std::vector<Real>()),
+    _intervals(getParam<std::vector<unsigned int>>("mesh_intervals")),
+    _region_ids(isParamValid("region_ids")
+                    ? getParam<std::vector<std::vector<subdomain_id_type>>>("region_ids")
+                    : std::vector<std::vector<subdomain_id_type>>()),
+    _extrude(getParam<bool>("extrude")),
+    _quad_center(getParam<bool>("quad_center_elements"))
+{
+  declareMeshProperty("pitch", _pitch);
+
+  // Ensure that the user has supplied the correct info for conformal mesh generation
+  if (getMeshByName(_reactor_params) != nullptr)
+    mooseError("The reactor_params mesh is not of the correct type");
+
+  if (!hasMeshProperty("mesh_dimensions", _reactor_params) ||
+      !hasMeshProperty("mesh_geometry", _reactor_params))
+    mooseError("The reactor_params input must be a ReactorMeshParams type MeshGenerator\n");
+  else
+  {
+    _mesh_dimensions = getMeshProperty<int>("mesh_dimensions", _reactor_params);
+    _mesh_geometry = getMeshProperty<std::string>("mesh_geometry", _reactor_params);
+  }
+  if (!_quad_center && _intervals[0] != 1)
+    mooseError("The number of mesh intervals in the ring region must be one if "
+               "quad_center_elements is not true");
+  if (_extrude && _mesh_dimensions != 3)
+    mooseError("This is a 2 dimensional mesh, you cannot extrude it. Check your ReactorMeshParams "
+               "inputs\n");
+  if (_extrude && (!hasMeshProperty("top_boundary_id", _reactor_params) ||
+                   !hasMeshProperty("bottom_boundary_id", _reactor_params)))
+    mooseError("Both top_boundary_id and bottom_boundary_id must be provided in ReactorMeshParams "
+               "if using extruded geometry");
+
+  if (_intervals.size() != (_ring_radii.size() + _duct_halfpitch.size() + 1))
+    mooseError("The number of mesh intervals must be equal to the number of annular regions + the "
+               "number of duct regions + 1"
+               " for the region between the rings and ducts\n");
+
+  if (isParamValid("region_ids"))
+  {
+    if (_region_ids[0].size() != (_ring_radii.size() + _duct_halfpitch.size() + 1))
+      mooseError("The number of region IDs given needs to be one more than the number of "
+                 "ring_radii + the number of duct_radii\n");
+  }
+  else
+  {
+    mooseError("Region IDs must be assigned with parameter region_ids");
+  }
```

Comment:
why not make it a requiredParam?

Reply:
How should this be done? Do we introduce another parameter like `define_optional_extra_ids`?

Reply:
No just make it non-required.
Then when retrieving it:
```
_region_ids(isParamValid("region_ids") ? getParam<...>("region_ids") : {} (or some empty vector of the right type)),
```

then when using it just check for the param to be set, or the vector to be non-empty

Reply:
The issue here is that `region_ids` is sort of required to set the block id's. We don't have a pathway where you can set block id's without defining region_ids

Reply:
You dont need to necessarily set block ids. 
If someone does not specify the 'region_ids' (soon to be block_ids) then you just use the default ones the MGs created?

Reply:
Should we have a quick meeting about the functionality of this system? I'm getting confused. region_id really should have been "material_id" in the first place. That was the original intent, but the author thought it would be good to use an intermediate name like 'region_id" instead of material_id just in case this caused a conflict somehow. Then, we thought, "why don't we just use the same integer for subdomain_id as we did for region_id (material_id)"? Hence the same number being used both places. Really, though, the subdomain_id doesn't need to be anything specific, it could be all 0s for tris and all 1s for quads. The individualized subdomain_ids is only ever really needed for mapping materials onto the mesh, which is already done at this point. Am I making sense? Should we discuss?

Reply:
I understand the region id is what you care the most about. But it needs to be clear that block ids are being modified.

It's just as important, if not more, for users of these meshes that are more likely to use blocks (an option available in way more moose objects than extra IDs) 

So either
- rename the parameter to block_ids
- add a message in region_ids parameter and in the 3 .md files saying that block_ids are being changed
- stop overwriting block ids with region ids

Reply:
I definitely want the block IDs to be predictable and matching them to region_ids makes the most sense EXCEPT in some scenarios where the same material is located in blocks of different types. This is the main conflict we know will happen especially as we add triangulated layers (coming up soon). We want to permit this. So, block_id won't always = region_id (material_id), and it shouldn't. Then, does your second option cover this case (a warning + documentation changes)?

Reply:
@GiudGiud one more question, could we have block_id = region_id (material_id) for all quad blocks, then do block_id = ${region_id}_TRI (a string) for the triangle blocks? I do not understand the interchangeability potential between ids and names at this point.

Reply:
You could set a block_name yes. The id will be different though. I'd be worried about conflicts if you dont intentionally set it, so you would still have to do that. 
Some mesh generators have two parameters, one for setting the id and one for the name. You could add that to these generators as well

Reply:
I'm hesitant to start making changes to this unless we can agree on what the best path forward is. Would it be enough to define the input parameter as `block_ids` and skip the process of copying over the block ID's to a separate extra integer? After thinking about this a bit more, users can use something like ExtraElementIDCopyGenerator to copy the subdomain id to a separate extra element integer if they really need this

Reply:
I don't think that is the way.

region_ids is needed downstream anyway so you'd need to add this ExtraElementIDCopyGenerator in the workflow, since the pin mixed-element-type issue is not obvious, and would require users to go deep into their mesh to merge each block through the region_ids. 
I think option 2, adding more info in parameters and the .md files (use !alert note format)  about the block IDs being modified is the easiest way forward

Reply:
We've gone ahead and implemented option 2 now. This still doesn't address the case where the user might have a situation where a shared material ID has both quad and tri elements (right now users are required to provide unique id's for region id's with quad-tri mixture elements which might otherwise share a material id), but in the future we are planning to have the following sort of convention:

All region id's will have a region_names parameter as well. Under the hood, RGMB will figure out whether the block has tri elements, and if so, will add the suffix "_TRI" to the region_names. This way, all block ID's are auto-generated by the underlying mesh generators, but users don't have to worry about this because they can just deal with block_names (based on region_names) or region_ids (same as material_ids) instead. Does this work?